### ISSUE TITLE
Change Request: support ESLINT_FORMATTER environment variable to set the default formatter

### ISSUE THREAD

**Issue by ThibautMarechal** [eslint/eslint#20946]
### ESLint version

v10

### What problem do you want to solve?

Currently the only way to use a custom formatter is via the `--format` CLI flag. There is no way to inject a formatter through the environment.

This creates a friction point CI setups where:

- Both CI & developers should use the same script: `my-package-manager run lint`.  
- CI agents need to use a specific formatter (e.g. for my case `eslint-formatter-teamcity`) to integrate build results natively into the CI platform, the developer only care about the default formatter. 
- I don't think that package.json scripts content should be aware of the environment where they run. 

While this was not exactly the same issue, this issue :[#19192](https://github.com/eslint/eslint/issues/19192)  asked for somethiong similar, but the config approach wasn't fitting the eslint architural choice, wich I agree. But my proposal could fix it too. 

### What do you think is the correct solution?

Add support for an `ESLINT_FORMATTER` environment variable, resolved at the CLI layer before output is produced:

```sh
# CI agent sets this once, globally
export ESLINT_FORMATTER=eslint-formatter-teamcity

# Script in package.json — unchanged, same as local dev
eslint .
```

Resolution order (suggested): `--format` flag > `ESLINT_FORMATTER` env var > default (`stylish`).

### Participation

- [x] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

I'm not sure how to deal with "how to resolve the formatter" dependending on the package manager.
The flat config help a lot becasue it rely on plain js import. But we cannot use that in this case. I could need some guidance here.

**Comment by DMartens** [comment]
Thank you for filing an issue for this.
I agree that this would be helpful also in cases where the team member wants a different formatter than the rest of the team.
I do not quite understand your question regarding the resolving logic. We can just reuse the [resolving logic](https://eslint.org/docs/latest/use/command-line-interface#-f---format) from `--format`. It would be possible that this would resolve to different versions of a formatter for different packages, but I think that would be fine.

**Comment by mdjermanovic** [comment]
I'm in favor of this proposal. However, since this introduces a new way to configure ESLint's behavior, while I don't think an RFC is necessary, I think this change still needs consensus from @eslint/eslint-tsc before being accepted.

**Comment by kecrily** [comment]
I still have a question about why we can’t specify the formatting tool in the configuration file. This doesn’t seem difficult to implement—what considerations led to the decision not to allow it to be specified in the configuration file?

**Comment by fasttime** [comment]
> ```sh
> # CI agent sets this once, globally
> export ESLINT_FORMATTER=eslint-formatter-teamcity
> 
> # Script in package.json — unchanged, same as local dev
> eslint .
> ```
> 
> Resolution order (suggested): `--format` flag > `ESLINT_FORMATTER` env var > default (`stylish`).

I think you could already achieve something like that today with:

```shell
eslint . --format ${ESLINT_FORMATTER:-stylish}
```

Of course, the syntax isn't portable, but setting environment variables isn't portable either.

**Comment by DMartens** [comment]
The TSC decided to reject this proposal as there are other ways to get this functionality, e.g. via the above example.

**Comment by ThibautMarechal** [comment]
The suggested workaround has real portability issues, it depends on shell-specific syntax which breaks across different environments (Windows CMD, PowerShell, bash, etc.), making it hard to share in team configs or CI scripts.

Would it be possible to expose this as a programmatic option or a config file setting instead? That would make it shell-agnostic and genuinely portable.

**Comment by DMartens** [comment]
There is no need for a programmatic option, as the Node.js API can already do this.
After linting the files, you can choose the formatter with `ESLint#loadFormatter`.
Example from the [documentation](https://eslint.org/docs/latest/integrate/nodejs-api#eslint-class):
```javascript
const { ESLint } = require("eslint");

(async function main() {
	// 1. Create an instance.
	const eslint = new ESLint();

	// 2. Lint files.
	const results = await eslint.lintFiles(["lib/**/*.js"]);

	// 3. Format the results.
	const formatter = await eslint.loadFormatter("stylish");
	const resultText = formatter.format(results);

	// 4. Output it.
	console.log(resultText);
})().catch(error => {
	process.exitCode = 1;
	console.error(error);
});
```

**Comment by ThibautMarechal** [comment]
If this is available programmatically, then I wonder why it's not directly available in the config file as proposed in the orriginal proposal (https://github.com/eslint/eslint/issues/19192). 

My point is, if a company want to use a specific formatter for all their repos in the CI, they can simply ship the formatter in the shared config file without creating a custom wrapper arround the eslint cli (meaning duplicating the example above multiple time per repo, or shipping a company wrapper) wich cause frustration to devs that just want to use standard and mainrstream cli and not company wrapper arround every tool they use.

**Comment by DMartens** [comment]
The reason why it is not part of the configuration file is given in the linked issue (it is CLI only).
You can also specify scripts in the `package.json` for each use case:
```json
{
	"scripts": {
		"lint": "eslint ."
		"CI": "npm run lint -- formatter='teamcity' && other_scripts_i_want_to_run"
	}
}
```

**Comment by ivangray271-lgtm** [comment]
Understood. Since the TSC has decided to keep this as a CLI-only concern
and the suggested package.json scripts or Node.js API are the recommended
paths forward, I will explore those alternatives for our CI setup.
